renku/cli/migrate.py

```diff
@@ -45,33 +48,84 @@ def migrate():
 @pass_local_client(
     clean=False,
     commit=True,
-    commit_only=COMMIT_DIFF_STRATEGY,
+    commit_only=DATASET_METADATA_PATHS + ['data/'],
 )
 @click.pass_context
 def datasets(ctx, client):
     """Migrate dataset metadata."""
-    from renku.models._jsonld import asjsonld
-    from renku.models.datasets import Dataset
-    from renku.models.refs import LinkReference
+    migrations = [
+        ensure_clean_lock,
+        migrate_datasets_pre_v0_3,
+        migrate_broken_paths,
+    ]
 
-    from ._checks.location_datasets import _dataset_metadata_pre_0_3_4
+    if all([migration(client) for migration in migrations]):
+        click.secho('OK', fg='green')
+
+
+def ensure_clean_lock(client):
+    """Make sure Renku lock file is not part of repository."""
+    lock_file = Path('.renku.lock')
+    if lock_file.exists():
+        lock_file.unlink()
 
-    for old_path in _dataset_metadata_pre_0_3_4(client):
-        dataset = Dataset.from_yaml(old_path, client=client)
 
+def migrate_datasets_pre_v0_3(client):
+    """Migrate datasets from Renku 0.3.x."""
+    for old_path in dataset_pre_0_3(client):
         name = str(old_path.parent.relative_to(client.path / 'data'))
+
+        dataset = Dataset.from_yaml(old_path, client=client)
         new_path = (client.renku_datasets_path / dataset.uid / client.METADATA)
         new_path.parent.mkdir(parents=True, exist_ok=True)
 
-        dataset = dataset.rename_files(
-            lambda key: os.path.
-            relpath(str(old_path.parent / key), start=str(new_path.parent))
-        )
+        with client.with_metadata() as meta:
+            for module in client.repo.submodules:
+                if Path(module.url).name == meta.name:
+                    module.remove()
 
-        with new_path.open('w') as fp:
-            yaml.dump(asjsonld(dataset), fp, default_flow_style=False)
-
-        old_path.unlink()
+        dataset.__reference__ = new_path
+        dataset.to_yaml()
 
+        Path(old_path).unlink()
         LinkReference.create(client=client,
                              name='datasets/' + name).set_reference(new_path)
+
+
+def migrate_broken_paths(client):
+    """Ensure all dataset and file paths are correct."""
+    for _, dataset in client.datasets.items():
+        dataset_path = Path(dataset.path)
+
+        expected_path = (client.renku_datasets_path / Path(dataset.identifier))
+
+        if not dataset_path.exists():
+            dataset_path = (
+                client.renku_datasets_path / uuid.UUID(dataset.identifier).hex
+            )
+
+        if not expected_path.exists():
+            shutil.move(str(dataset_path), str(expected_path))
+            dataset.path = expected_path
+            dataset.__reference__ = expected_path / client.METADATA
+
+        for file_ in dataset.files:
```

Comment:
check here whether files have paths outside the repository root - if so, renormalize the paths

Reply:
This check is right here, 2 line below. Not sure what are you referring to.